File: vol2/ch21/LeetCode32.py
class Solution:
  def longestValidParentheses(self,s:str) -> int:
    n=len(s)
    dp=[0]*n      #一维动态规划数组
    ans=0
    for i in range(1,n):
      if s[i]==')':
        if s[i-1]=='(':
          if i>=2:dp[i]=dp[i-2]+2
          else:dp[i]=2
        elif i-dp[i-1]>0 and s[i-dp[i-1]-1]=='(':
          if i-dp[i-1]>=2:
            dp[i]=dp[i-1]+dp[i-dp[i-1]-2]+2
          else:
            dp[i]=dp[i-1]+2
      ans=max(ans,dp[i])
    return ans

File: vol2/ch21/LeetCode32.cpp
class Solution {
public:
  int longestValidParentheses(string s) {
    int n=s.size();
    vector<int> dp(n,0);              //一维动态规划数组
    int ans=0;
    for(int i=1;i<n;i++) {
      if (s[i]==')') {
        if (s[i-1]=='(') {
           if(i>=2) dp[i]=dp[i-2]+2;
           else dp[i]=2;
          }
          else if(i-dp[i-1]>0 && s[i-dp[i-1]-1]=='(') {
            if(i-dp[i-1]>=2) dp[i]=dp[i-1]+dp[i-dp[i-1]-2]+2;
            else dp[i]=dp[i-1]+2;
          }
      }
      ans=max(ans,dp[i]);
    }
    return ans;
  }
};

